conditionally add request metadata length to log output (no test)

smokytheangel0 opened the pull request:
This pull request implements the first step in determining a maximum metadata size for uploads as described in issue 1896 [https://github.com/rust-lang/crates.io/issues/1896](url). There was no test added for this change as log output is not automatically tested.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @carols10cents (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/crates.io/blob/master/docs/CONTRIBUTING.md) for more information.

smokytheangel0 commented:
waiting on a full local build to manually test the addition.

carols10cents commented on src/middleware/log_request.rs:
```diff
@@ -31,19 +31,37 @@ impl Handler for LogRequests {
-            request_id = request_header(req, "X-Request-Id"),
-            status = response_code,
-        );
+        if let Some(metadata_length) = req.extensions().find::<u64>() {
```
This adds a lot of duplication. What do you think about, before the `print`, doing something like:

```
let metadata_length = req.extensions().find::<u64>().map_or(String::new(), |n| format!(" metadata_length={}", n));
```

which will set `metadata_length` to an empty string if there isn't a length, and will set `metadata_length` to ` metadata_length=32` if the length is `Some(32)`.

Then the `print` can unconditionally print out the value in `metadata_length`.

Does that make sense? If not, please ask questions!

smokytheangel0 replied:
this makes perfect sense, I was wondering myself how to reduce the duplicity of the branches, and I am not very familiar with map/reduce type things, so this will be an excellent change

smokytheangel0 commented:
I think I may have handled git wrong here, I reverted some other changes and added crates.io upstream so that I could stay level with it in the future, I don't see anything wrong with the latest commit though, so please let me know if I messed anything up.

carols10cents commented on src/middleware/log_request.rs:
```diff
@@ -31,10 +31,16 @@ impl Handler for LogRequests {
              request_id={request_id} fwd=\"{ip}\" service={time_ms}ms \
-             status={status} user_agent=\"{user_agent}\"",
+             status={status} user_agent=\"{user_agent}\"\
+             metadata_length={metadata_length}",
```
When I test this out locally, I'm seeing this on requests that don't have a `metadata_length`:

```
at=info method=GET path="/api/v1/summary" request_id= fwd="" service=44ms status=200 user_agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:72.0) Gecko/20100101 Firefox/72.0"metadata_length=
```

And this on requests that do:

```
at=info method=PUT path="/api/v1/crates/new" request_id= fwd="" service=1060ms status=200 user_agent="cargo 1.39.0 (1c6ec66d5 2019-09-30)"metadata_length= metadata_length=734
```

I think we can remove the `metadata_length=` here now because it's duplicated with [this](https://github.com/rust-lang/crates.io/pull/1905/files#diff-3c16efcf7a504a2aae66b1d607874710R37) ` metadata_length=`.

Could you make that change please? Once that's done, I think this is good to go! And don't worry about the commits, I can fix them up 👍

smokytheangel0 replied:
I definitely should have noticed that, thanks.

carols10cents commented:
Thank you so much!!!

@bors r+

bors commented:
:pushpin: Commit 92c4dcb64b71484d7a387a0d926134f9395ec20a has been approved by `carols10cents`

<!-- @bors r=carols10cents 92c4dcb64b71484d7a387a0d926134f9395ec20a -->
<!-- homu: {"type":"Approved","sha":"92c4dcb64b71484d7a387a0d926134f9395ec20a","approver":"carols10cents"} -->

bors commented:
:hourglass: Testing commit 92c4dcb64b71484d7a387a0d926134f9395ec20a with merge 1c98d073cb6d3181fd74bad9706010f37c810ce3...
<!-- homu: {"type":"BuildStarted","head_sha":"92c4dcb64b71484d7a387a0d926134f9395ec20a","merge_sha":"1c98d073cb6d3181fd74bad9706010f37c810ce3"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/crates.io/builds/138061362)
Approved by: carols10cents
Pushing 1c98d073cb6d3181fd74bad9706010f37c810ce3 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"carols10cents","base_ref":"master","builders":{"checks-travis":"https://travis-ci.com/rust-lang/crates.io/builds/138061362"},"merge_sha":"1c98d073cb6d3181fd74bad9706010f37c810ce3"} -->
